File: core/src/feedbacker_core/originals.py
"""Import sampled original files from bulk downloads and extract their text (#15).

A sample may be spread across several sources (e.g. a main zip and a zip from
a late-submission point, or single files). Only the sampled submissions are
taken; other students' files are never opened. Each source used is copied into
the workspace so it is deleted with it. Real file names go only into the pseudonym key. Each submission is
imported completely or not at all, and failures are listed.
"""

from __future__ import annotations

import hashlib
from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path

from feedbacker_core.archive import SUPPORTED, select_members
from feedbacker_core.extract import ExtractionError, extract, source_format
from feedbacker_core.models import (
    Actor,
    ActorKind,
    Provenance,
    SourceKind,
    Submission,
    Transformation,
)
from feedbacker_core.request import load_request
from feedbacker_core.workspace import PseudonymKey, Workspace, WorkspaceError

MODERATOR = Actor(kind=ActorKind.MODERATOR, label="moderator")
SOURCES = "sources"
SUBMISSIONS = "submissions"


class ImportProblem(ValueError):
    def __init__(self, problems: list[str]):
        self.problems = problems
        super().__init__("cannot import originals:\n- " + "\n- ".join(problems))


@dataclass
class ImportResult:
    imported: list[Submission] = field(default_factory=list)
    failed: dict[str, str] = field(default_factory=dict)  # submission_id -> reason
    ignored_count: int = 0


def submission_path(submission_id: str) -> str:
    return f"{SUBMISSIONS}/{submission_id}.json"


def import_originals(
    workspace: Workspace,
    sources: Path | list[Path],
    *,
    replace: bool = False,
    now: datetime | None = None,
) -> ImportResult:
    """Import the sampled originals found across ``sources`` (zips or single files)."""
    if isinstance(sources, Path):
        sources = [sources]
    request = load_request(workspace)
    key = workspace.read_key()
    sampled = []
    for s in request.sample:
        entry = key.by_pseudonym(s.pseudonym)
        assert entry is not None  # load_request guarantees the key resolves
        sampled.append((s, entry))

    existing = [
        s.submission_id for s, _ in sampled if workspace.exists(submission_path(s.submission_id))
    ]
    if existing and not replace:
        raise WorkspaceError(
            f"submissions already imported ({', '.join(existing)}); use replace to import again"
        )

    try:
        selection = select_members(sources, [e.external_id for _, e in sampled])
    except ValueError as err:
        raise ImportProblem([str(err)]) from None
    problems = list(selection.problems)
    for external_id, member in selection.matched.items():
        if member.suffix not in SUPPORTED:
            problems.append(
                f"file for identifier '{external_id}' is '{member.suffix or 'no type'}'; "
                "Stage 0 imports typed docx and pdf only"
            )
    if problems:
        raise ImportProblem(problems)

    timestamp = now or datetime.now(UTC)
    source_hashes = {
        src: hashlib.sha256(src.read_bytes()).hexdigest()
        for src in {m.source for m in selection.matched.values()}
    }

    result = ImportResult(ignored_count=selection.ignored_count)
    new_entries = []
    for s, entry in sampled:
        member = selection.matched[entry.external_id]
        data = member.read()
        # Only the selected file is stored, never the whole bulk download.
        stored = workspace.path / SOURCES / "originals" / f"{s.submission_id}{member.suffix}"
        stored.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
        stored.write_bytes(data)
        stored.chmod(0o600)
        digest = hashlib.sha256(data).hexdigest()
        source_hash = source_hashes[member.source]
        try:
            extracted = extract(stored, now=timestamp)
        except ExtractionError as err:
            result.failed[s.submission_id] = str(err)
            stored.unlink(missing_ok=True)
            new_entries.append(entry)
            continue
        kind = "archive" if member.is_archive else "file"
        submission = Submission(
            id=s.submission_id,
            pseudonym=s.pseudonym,
            source_kind=SourceKind.ORIGINAL,
            source_format=source_format(stored),
            source_sha256=digest,
            extract=extracted,
            provenance=Provenance(
                source=f"{kind}:sha256:{source_hash}",
                transformation=Transformation.IMPORTED,
                actor=MODERATOR,
                timestamp=timestamp,
                input_hashes=sorted({source_hash, digest}),
            ),
        )
        new_entries.append(
            entry.model_copy(
                update={"source_files": {**entry.source_files, "original": member.file_name}}
            )
        )
        result.imported.append(submission)

    # Key first (it only gains information), then the submission records.
    untouched = [e for e in key.entries if e.pseudonym not in {s.pseudonym for s, _ in sampled}]
    workspace.write_key(PseudonymKey(entries=_in_key_order(key, untouched + new_entries)))
    for submission in result.imported:
        workspace.write_json(
            submission_path(submission.id), submission.model_dump(mode="json"), private=True
        )
    return result


def _in_key_order(key: PseudonymKey, entries: list) -> list:
    order = {e.pseudonym: i for i, e in enumerate(key.entries)}
    return sorted(entries, key=lambda e: order[e.pseudonym])


def load_submission(workspace: Workspace, submission_id: str) -> Submission:
    path = submission_path(submission_id)
    if not workspace.exists(path):
        raise WorkspaceError(f"submission {submission_id} has not been imported")
    return Submission.model_validate(workspace.read_json(path))

File: ui/src/core/originals.ts
/**
 * Import sampled original files from bulk downloads and extract their text: a
 * port of `core/src/feedbacker_core/originals.py` (#15, #48).
 *
 * A sample may be spread across several sources (a main zip, a zip from a
 * late-submission point, single files). Only the sampled submissions are
 * taken; other students' files are never opened. Bulk downloads are never
 * copied into the workspace: only each selected file is stored, under its
 * pseudonymous submission ID, and only hashes of the downloads are recorded.
 * Real file names go only into the pseudonym key. Each submission is imported
 * completely or not at all, and failures are listed.
 *
 * Python stages each file in a `.staging` folder so that nothing existing is
 * touched until a submission has succeeded; here the staged bytes are held
 * in memory and written only after every submission has been processed,
 * which gives the same guarantee without a staging folder in the workspace.
 */

import { ArchiveError, selectMembers } from "./archive.ts";
import { extract, ExtractionError, sha256Bytes, sourceFormat } from "./extract.ts";
import { Submission } from "./models.ts";
import { loadRequest, MODERATOR } from "./request.ts";
import { byPseudonym, withEntries, type KeyEntry, type PseudonymKey, type Workspace, WorkspaceError } from "./workspace.ts";
import { hashSource, type ByteSource } from "./zip.ts";

export const SOURCES = "sources";
export const SUBMISSIONS = "submissions";
const ORIGINALS = `${SOURCES}/originals`;
const SUPPORTED = new Set([".docx", ".pdf"]);

export class ImportProblem extends Error {
  readonly problems: string[];

  constructor(problems: string[]) {
    super("cannot import originals:\n- " + problems.join("\n- "));
    this.name = "ImportProblem";
    this.problems = problems;
  }
}

export interface ImportResult {
  imported: Submission[];
  failed: Map<string, string>; // submission_id -> reason
  ignoredCount: number;
}

export const submissionPath = (submissionId: string) => `${SUBMISSIONS}/${submissionId}.json`;

/** Import the sampled originals found across `sources` (zips or single files). */
export async function importOriginals(
  ws: Workspace,
  sources: ByteSource | ByteSource[],
  options: { replace?: boolean; now?: Date } = {},
): Promise<ImportResult> {
  const list = Array.isArray(sources) ? sources : [sources];
  const request = await loadRequest(ws);
  const key = await ws.readKey();
  const sampled = request.sample.map((s) => [s, byPseudonym(key, s.pseudonym)!] as const); // loadRequest guarantees these resolve

  const existing: string[] = [];
  for (const [s] of sampled) if (await ws.exists(submissionPath(s.submission_id))) existing.push(s.submission_id);
  if (existing.length && !options.replace) {
    throw new WorkspaceError(`submissions already imported (${existing.join(", ")}); use replace to import again`);
  }

  let selection;
  try {
    selection = await selectMembers(list, sampled.map(([, e]) => e.external_id));
  } catch (err) {
    if (err instanceof ArchiveError) throw new ImportProblem([err.message]);
    throw err;
  }
  const labels = new Map(sampled.map(([s, e]) => [e.external_id, `${s.pseudonym} (${s.submission_id})`]));
  const problems = selection.problems((id) => labels.get(id)!, list);
  for (const [externalId, member] of selection.matched) {
    if (!SUPPORTED.has(member.suffix)) {
      problems.push(`file for ${labels.get(externalId)} is not docx or pdf; Stage 0 imports typed docx and pdf only`);
    }
  }
  if (problems.length) throw new ImportProblem(problems);

  const now = options.now ?? new Date();
  const sourceHashes = new Map<ByteSource, string>();
  for (const member of selection.matched.values()) {
    if (!sourceHashes.has(member.source)) sourceHashes.set(member.source, await hashSource(member.source));
  }

  const result: ImportResult = { imported: [], failed: new Map(), ignoredCount: selection.ignoredCount };
  const newEntries: KeyEntry[] = [];
  const staged: { submission: Submission; bytes: Uint8Array; path: string }[] = [];
  for (const [s, entry] of sampled) {
    const member = selection.matched.get(entry.external_id)!;
    const fileName = `${s.submission_id}${member.suffix}`;
    // Nothing existing is touched until every submission has been processed, so
    // a failed replacement leaves the previous file and record intact.
    let bytes: Uint8Array;
    try {
      bytes = await member.read();
    } catch (err) {
      result.failed.set(s.submission_id, `the selected file could not be read (${(err as Error).name})`);
      newEntries.push(entry);
      continue;
    }
    const digest = sha256Bytes(bytes);
    const sourceHash = sourceHashes.get(member.source)!;
    let extracted;
    try {
      extracted = await extract(fileName, bytes, now);
    } catch (err) {
      if (!(err instanceof ExtractionError)) throw err;
      result.failed.set(s.submission_id, err.message);
      newEntries.push(entry);
      continue;
    }
    const submission = Submission.parse({
      id: s.submission_id,
      pseudonym: s.pseudonym,
      source_kind: "original",
      source_format: sourceFormat(fileName),
      source_sha256: digest,
      extract: extracted,
      provenance: {
        source: `${member.isArchive ? "archive" : "file"}:sha256:${sourceHash}`,
        transformation: "imported",
        actor: MODERATOR,
        timestamp: now.toISOString(),
        input_hashes: [...new Set([sourceHash, digest])].sort(),
      },
    });
    newEntries.push({ ...entry, source_files: { ...entry.source_files, original: member.fileName } });
    staged.push({ submission, bytes, path: `${ORIGINALS}/${fileName}` });
    result.imported.push(submission);
  }

  // Key first (it only gains information); then, per submission, the source
  // file and its record. Each write is atomic, and they are ordered so that
  // either the previous pair or the new one is always complete: an older
  // original in another format is removed only once its replacement and
  // record are written, and a same-named original is restored if its record
  // can't be written. loadSubmission detects any mismatch that remains.
  const sampledPseudonyms = new Set(sampled.map(([s]) => s.pseudonym));
  const untouched = key.entries.filter((e) => !sampledPseudonyms.has(e.pseudonym));
  await ws.writeKey(withEntries(key, inKeyOrder(key, [...untouched, ...newEntries])));
  const present = (await ws.exists(ORIGINALS)) ? await ws.fs.list(ORIGINALS) : [];
  let written = false;
  try {
    for (const { submission, bytes, path } of staged) {
      const previous = await ws.readBytes(path);
      written = true;
      await ws.writeBytes(path, bytes);
      try {
        await ws.writeJson(submissionPath(submission.id), submission);
      } catch (err) {
        await (previous ? ws.writeBytes(path, previous) : ws.fs.remove(path)).catch(() => {});
        throw err;
      }
      for (const old of present) {
        const oldPath = `${ORIGINALS}/${old.name}`;
        if (old.kind === "file" && old.name.startsWith(`${submission.id}.`) && oldPath !== path) await ws.fs.remove(oldPath);
      }
    }
  } catch (err) {
    // Whatever was written before the failure is still made private.
    if (written) await ws.secure().catch(() => {});
    throw err;
  }
  if (written) await ws.secure(); // stored originals and records are private
  return result;
}

function inKeyOrder(key: PseudonymKey, entries: KeyEntry[]): KeyEntry[] {
  const order = new Map(key.entries.map((e, i) => [e.pseudonym, i]));
  return [...entries].sort((a, b) => order.get(a.pseudonym)! - order.get(b.pseudonym)!);
}

/** Load a submission and confirm its stored source file still matches the record. */
export async function loadSubmission(ws: Workspace, submissionId: string): Promise<Submission> {
  const path = submissionPath(submissionId);
  if (!(await ws.exists(path))) throw new WorkspaceError(`submission ${submissionId} has not been imported`);
  const parsed = Submission.safeParse(await ws.readJson(path));
  if (!parsed.success) throw new WorkspaceError(`submission ${submissionId}: its record is not valid`);
  const stored = await ws.readBytes(`${ORIGINALS}/${submissionId}.${parsed.data.source_format}`);
  if (!stored || sha256Bytes(stored) !== parsed.data.source_sha256) {
    throw new WorkspaceError(
      `submission ${submissionId}: its stored source file is missing or does not match ` +
        "the record; the workspace may be damaged, so import it again",
    );
  }
  return parsed.data;
}
